### ISSUE TITLE
Missing documentation: awslogs doesn't support AWS credentials files contrary to documentation

### ISSUE THREAD

**Issue by cppexpert** [docker/docker#19392]
It looks like awslogs doesn't support AWS credentials files contrary to [documentation](https://docs.docker.com/engine/reference/logging/awslogs/). The documentation says

```
You can provide these credentials with the AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY, and AWS_SESSION_TOKEN environment variables, the default AWS shared credentials file (~/.aws/credentials of the root user)
```

First of all providing environment variables **and** credentials file is redundant because it caries same information. Secondly, hours of debugging indicate that Docker doesn't open `/root/.aws/credentials` at all which was verified with watching file events using `fatrace`.

As result `docker run` always fails with

```
Error response from daemon: Cannot start container case: Failed to initialize logging driver: NoCredentialProviders: no valid providers in chain
```

In order to make sure it's not me going crazy I wrote following simple application in Scala to doublecheck that AWS SDK does indeed works with credentials file

```scala
val conf = ConfigFactory.load().getConfig("case")
val aws_region = RegionUtils.getRegion(conf.getString("aws_region"))
val client = new AWSLogsClient(new ProfileCredentialsProvider("default"))
client.setRegion(aws_region)
val req = new PutLogEventsRequest()
req.setLogGroupName("case")
req.setLogStreamName("api-server")
val arr: util.Collection[InputLogEvent] = new util.ArrayList[InputLogEvent]()
val event = new InputLogEvent()
event.setMessage("test message!")
event.setTimestamp(ZonedDateTime.now().toEpochSecond * 1000)
arr.add(event)
req.setLogEvents(arr)
val regs = client.putLogEvents(req)
println(regs)
```
Could you confirm that support of credentials file was removed in 1.9.1 or explain how to make it work ?

Thanks!

---

Containers: 6
Images: 83
Server Version: 1.9.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 95
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.2.0-23-generic
Operating System: Ubuntu 15.10
CPUs: 8
Total Memory: 15.61 GiB
Name: expert-x220
ID: 3CN5:AOOV:DKXE:5WUH:HFTZ:2744:MU7U:REWH:LFPQ:M5K4:3SM5:WMXA
Username: expert
Registry: https://index.docker.io/v1/
WARNING: No swap limit support

Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:20:08 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:20:08 UTC 2015
 OS/Arch:      linux/amd64

**Comment by thaJeztah** [comment]
I think these are used by default;

https://github.com/docker/docker/blob/c545dce004a4b833994478a2fc993ee8e2b3b2cd/vendor/src/github.com/aws/aws-sdk-go/aws/defaults/defaults.go#L18-L23

https://github.com/docker/docker/blob/c545dce004a4b833994478a2fc993ee8e2b3b2cd/vendor/src/github.com/aws/aws-sdk-go/aws/credentials/shared_credentials_provider.go#L24-L40


But I'm not on AWS, so perhaps @samuelkarp can check

**Comment by cppexpert** [comment]
I just tried adding `AWS_SHARED_CREDENTIALS_FILE` to my systemd script. Unfortunately it didn't help :(

```
[Unit]
Description=Docker Application Container Engine
Documentation=https://docs.docker.com
After=network.target docker.socket
Requires=docker.socket

[Service]
Environment="AWS_SHARED_CREDENTIALS_FILE=/home/expert/.aws/credentials"
Type=notify
ExecStart=/usr/bin/docker daemon -H fd://
MountFlags=slave
LimitNOFILE=1048576
LimitNPROC=1048576
LimitCORE=infinity

[Install]
WantedBy=multi-user.target
```

**Comment by HackToday** [comment]
Did it work in prior 1.9.1 ? 
from Failed to initialize logging driver: NoCredentialProviders: no valid providers in chain
it seems that docker load related providers, I think seems docker daemon need configure that provide to make docker load aws related helper

**Comment by cppexpert** [comment]
Any updates, guys?

Thanks.

**Comment by samuelkarp** [comment]
@cppexpert, thanks for reporting this.  Depending on how the Docker daemon is started (and possibly the init system used), there may be no `HOME` set when the daemon starts.  I was able to define `HOME` for myself, and then the credentials file in `$HOME/.aws/credentials` was properly read.  It sounds like there's room to improve the documentation; if you have specific suggestions I'd like to hear them.

**Comment by samuelkarp** [comment]
@cppexpert The reason setting `AWS_SHARED_CREDENTIALS_FILE` did not have any effect is that Docker 1.9 uses version 0.7.1 of the AWS SDK for Go, which [does not have support for `AWS_SHARED_CREDENTIALS_FILE`](https://github.com/aws/aws-sdk-go/blob/v0.7.1/aws/credentials/shared_credentials_provider.go#L107-L121).  Docker 1.10 has a newer version of the SDK (v0.9.9) which adds that support, so Docker 1.10 should be able to use `AWS_SHARED_CREDENTIALS_FILE` to find the credentials.

**Comment by aditirajagopal** [comment]
@cppexpert can you verify that this works for Docker 1.10 so this issue can be closed?

**Comment by cppexpert** [comment]
I'll be able to check on Monday if that's ok.

**Comment by ewales** [comment]
Running Docker 1.10 and it does not work: 

`root@ewales-ubuntu:~# docker info
Containers: 56
 Running: 4
 Paused: 0
 Stopped: 52
Images: 641
Server Version: 1.10.0
Storage Driver: devicemapper
 Pool Name: docker-8:18-7609062-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 107.4 GB
 Backing Filesystem: ext4
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 25.36 GB
 Data Space Total: 107.4 GB
 Data Space Available: 82.01 GB
 Metadata Space Used: 53.38 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.094 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.99 (2015-06-20)
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: null host bridge
Kernel Version: 4.2.0-27-generic
Operating System: Ubuntu 15.10
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 15.62 GiB
Name: ewales-ubuntu
ID: 7375:OBEQ:KZB6:V4LS:KRWZ:7CQ7:5RVV:ISJZ:O5MP:3B4R:D4EO:VQCD
Username: ewales
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
root@ewales-ubuntu:~# docker --version
Docker version 1.10.0, build 590d5108
root@ewales-ubuntu:~# docker run -d -P --log-driver=awslogs --log-opt awslogs-region=us-east-1 --log-opt awslogs-group=test/test.log --log-opt awslogs-stream=ewales redis
e68ea42649240c606803b0ce2532ed5c8498c1fcbad4bd77a51cc446bbc42109
docker: Error response from daemon: Failed to initialize logging driver: NoCredentialProviders: no valid providers in chain.
root@ewales-ubuntu:~# ls -la $AWS_SHARED_CREDENTIALS_FILE
-rw------- 1 root root 116 feb 23 08:01 /root/.aws/credentials
root@ewales-ubuntu:~#`

**Comment by tnir** [comment]
It does not work for me, on Docker Engine 1.11.1, launched by `docker-machine create --driver amazonec2`

```
$ docker info
Containers: x
 Running: x
 Paused: x
 Stopped: x
Images: x
Server Version: 1.11.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: xx
 Dirperm1 Supported: true
Logging Driver: json-file
Plugins:
 Volume: local
 Network: host bridge null
Kernel Version: 4.2.0-36-generic
Operating System: Ubuntu 15.10
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 991.1 MiB
Name: xxxxxxxxxxxxxxxxxxxx
ID: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
WARNING: No swap limit support
Labels:
 provider=amazonec2
```

**Comment by samuelkarp** [comment]
On a computer with Docker 1.11.1, I was able to use `AWS_SHARED_CREDENTIALS_FILE` by doing this:

```bash
$ echo 'export AWS_SHARED_CREDENTIALS_FILE=/home/sam/.aws/credentials' | sudo tee -a /etc/default/docker
$ sudo service docker stop && sudo service docker start
```

@ewales @tnir Are you sure the environment variable is set for the Docker daemon?

**Comment by ewales** [comment]
@samuelkarp  Can confirm it works as you describe with Docker 1.11.1. I feel silly for not realizing the daemon needed the variable. Thanks!

**Comment by thaJeztah** [comment]
Looks like we can close this issue then, thanks!

**Comment by ottumm** [comment]
Is there a way to set this environment variable for the docker daemon when using Docker for Mac?

**Comment by TexasDude071** [comment]
I was able to use AWS_SHARED_CREDENTIALS_FILE on Docker 1.10.3 under CoreOS 1068.10.0.  From the documentation, I expected to put a [us-east-1] section in the credentials file, but, that left me with the NoCredentialsProvided error.  Once I replaced the [us-east-1] with [Default] as the section heading then, everything seemed to work. I do set awslogs-region=us-east-1 in the Docker systemd unit file.  

Thinking through this, it makes me wonder if I should set awslogs-region=us-east and create a 'config' file next to the 'credentials' file that expands on how 'us-east' should be treated.  This would be consistent with how I've used awscli.  Not sure, but, in any event, I"m glad to now have my logs in Cloudwatch.
